File: Firebase/Database/Snapshot/FIndexedNode.m
/*
 * Copyright 2017 Google
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#import "FIndexedNode.h"

#import "FImmutableSortedSet.h"
#import "FIndex.h"
#import "FPriorityIndex.h"
#import "FKeyIndex.h"
#import "FChildrenNode.h"

static FImmutableSortedSet *FALLBACK_INDEX;

@interface FIndexedNode ()

@property (nonatomic, strong) id<FNode> node;
/**
 * The indexed set is initialized lazily to prevent creation when it is not needed
 */
@property (nonatomic, strong) FImmutableSortedSet *indexed;
@property (nonatomic, strong) id<FIndex> index;

@end

@implementation FIndexedNode

+ (FImmutableSortedSet *)fallbackIndex {
    static FImmutableSortedSet *fallbackIndex;
    static dispatch_once_t once;
    dispatch_once(&once, ^{
        fallbackIndex = [[FImmutableSortedSet alloc] init];
    });
    return fallbackIndex;
}

+ (FIndexedNode *)indexedNodeWithNode:(id<FNode>)node
{
    return [[FIndexedNode alloc] initWithNode:node index:[FPriorityIndex priorityIndex]];
}

+ (FIndexedNode *)indexedNodeWithNode:(id<FNode>)node index:(id<FIndex>)index
{
    return [[FIndexedNode alloc] initWithNode:node index:index];
}

- (id)initWithNode:(id<FNode>)node index:(id<FIndex>)index
{
    // Initialize indexed lazily
    return [self initWithNode:node index:index indexed:nil];
}

- (id)initWithNode:(id<FNode>)node index:(id<FIndex>)index indexed:(FImmutableSortedSet *)indexed
{
    self = [super init];
    if (self != nil) {
        self->_node = node;
        self->_index = index;
        self->_indexed = indexed;
    }
    return self;
}

- (void)ensureIndexed
{
    if (!self.indexed) {
        if ([self.index isEqual:[FKeyIndex keyIndex]]) {
            self.indexed = [FIndexedNode fallbackIndex];
        } else {
            __block BOOL sawChild;
            [self.node enumerateChildrenUsingBlock:^(NSString *key, id<FNode> node, BOOL *stop) {
                sawChild = sawChild || [self.index isDefinedOn:node];
                *stop = sawChild;
            }];
            if (sawChild) {
                NSMutableDictionary *dict = [NSMutableDictionary dictionary];
                [self.node enumerateChildrenUsingBlock:^(NSString *key, id<FNode> node, BOOL *stop) {
                    FNamedNode *namedNode = [[FNamedNode alloc] initWithName:key andNode:node];
                    dict[namedNode] = [NSNull null];
                }];
                // Make sure to assign index here, because the comparator will be retained and using self will cause a
                // cycle
                id<FIndex> index = self.index;
                self.indexed = [FImmutableSortedSet setWithKeysFromDictionary:dict
                                                                   comparator:^NSComparisonResult(FNamedNode *namedNode1, FNamedNode *namedNode2) {
                    return [index compareNamedNode:namedNode1 toNamedNode:namedNode2];
                }];
            } else {
                self.indexed = [FIndexedNode fallbackIndex];
            }
        }
    }
}

- (BOOL)hasIndex:(id<FIndex>)index
{
    return [self.index isEqual:index];
}

- (FIndexedNode *)updateChild:(NSString *)key withNewChild:(id<FNode>)newChildNode
{
    id<FNode> newNode = [self.node updateImmediateChild:key withNewChild:newChildNode];
    if (self.indexed == [FIndexedNode fallbackIndex] && ![self.index isDefinedOn:newChildNode]) {
        // doesn't affect the index, no need to create an index
        return [[FIndexedNode alloc] initWithNode:newNode index:self.index indexed:[FIndexedNode fallbackIndex]];
    } else if (!self.indexed || self.indexed == [FIndexedNode fallbackIndex]) {
        // No need to index yet, index lazily
        return [[FIndexedNode alloc] initWithNode:newNode index:self.index];
    } else {
        id<FNode> oldChild = [self.node getImmediateChild:key];
        FImmutableSortedSet *newIndexed = [self.indexed removeObject:[FNamedNode nodeWithName:key node:oldChild]];
        if (![newChildNode isEmpty]) {
            newIndexed = [newIndexed addObject:[FNamedNode nodeWithName:key node:newChildNode]];
        }
        return [[FIndexedNode alloc] initWithNode:newNode index:self.index indexed:newIndexed];
    }
}

- (FIndexedNode *)updatePriority:(id<FNode>)priority
{
    return [[FIndexedNode alloc] initWithNode:[self.node updatePriority:priority]
                                        index:self.index
                                      indexed:self.indexed];
}

- (FNamedNode *)firstChild
{
    if (![self.node isKindOfClass:[FChildrenNode class]]) {
        return nil;
    } else {
        [self ensureIndexed];
        if (self.indexed == [FIndexedNode fallbackIndex]) {
            return [((FChildrenNode *)self.node) firstChild];
        } else {
            return self.indexed.firstObject;
        }
    }
}

- (FNamedNode *)lastChild
{
    if (![self.node isKindOfClass:[FChildrenNode class]]) {
        return nil;
    } else {
        [self ensureIndexed];
        if (self.indexed == [FIndexedNode fallbackIndex]) {
            return [((FChildrenNode *)self.node) lastChild];
        } else {
            return self.indexed.lastObject;
        }
    }
}

- (NSString *)predecessorForChildKey:(NSString *)childKey childNode:(id<FNode>)childNode index:(id<FIndex>)index
{
    if (![self.index isEqual:index]) {
        [NSException raise:NSInvalidArgumentException format:@"Index not available in IndexedNode!"];
    }
    [self ensureIndexed];
    if (self.indexed == [FIndexedNode fallbackIndex]) {
        return [self.node predecessorChildKey:childKey];
    } else {
        FNamedNode *node = [self.indexed predecessorEntry:[FNamedNode nodeWithName:childKey node:childNode]];
        return node.name;
    }
}

- (void)enumerateChildrenReverse:(BOOL)reverse usingBlock:(void (^)(NSString *, id<FNode>, BOOL *))block
{
    [self ensureIndexed];
    if (self.indexed == [FIndexedNode fallbackIndex]) {
        [self.node enumerateChildrenReverse:reverse usingBlock:block];
    } else {
        [self.indexed enumerateObjectsReverse:reverse usingBlock:^(FNamedNode *namedNode, BOOL *stop) {
            block(namedNode.name, namedNode.node, stop);
        }];
    }
}

- (NSEnumerator *)childEnumerator
{
    [self ensureIndexed];
    if (self.indexed == [FIndexedNode fallbackIndex]) {
        return [self.node childEnumerator];
    } else {
        return [self.indexed objectEnumerator];
    }
}

@end

File: FirebaseDatabaseSwiftCore/Sources/FIndexedNode.swift
//
//  File.swift
//  
//
//  Created by Morten Bek Ditlevsen on 27/09/2021.
//

import SortedCollections
import Foundation

/**
 * Represents a node together with an index. The index and node are updated in
 * unison. In the case where the index does not affect the ordering (i.e. the
 * ordering is identical to the key ordering) this class uses a fallback index
 * to save memory. Everything operating on the index must special case the
 * fallback index.
 */

public struct FIndexedNode {
    struct IndexedNamedNode: Comparable, Equatable {
        static func == (lhs: IndexedNamedNode, rhs: IndexedNamedNode) -> Bool {
            lhs.wrapped == rhs.wrapped && lhs.index === rhs.index
        }

        static func < (lhs: IndexedNamedNode, rhs: IndexedNamedNode) -> Bool {
            lhs.index.compareNamedNode(lhs.wrapped, toNamedNode: rhs.wrapped) == .orderedAscending
        }

        let wrapped: FNamedNode
        let index: FIndex
    }

    enum Wrapper {
        case fallback
        case indexed(SortedSet<IndexedNamedNode>)
    }

    let node: FNode
    let index: FIndex
    var indexed: Wrapper {
        if let initial = initialIndexed {
            return initial
        }
        if index.isEqual(FKeyIndex.keyIndex) {
            return .fallback
        }
        var sawChild = false
        node.enumerateChildren { key, node, stop in
            sawChild = sawChild || self.index.isDefined(on: node)
            if sawChild {
                stop.pointee = ObjCBool(booleanLiteral: true)
            }
        }
        if sawChild {
            var set: SortedSet<IndexedNamedNode> = []
            node.enumerateChildren { [index] key, node, stop in
                let namedNode = FNamedNode(name: key, andNode: node)
                set.insert(IndexedNamedNode(wrapped: namedNode, index: index))
            }

            return .indexed(set)
        } else {
            return .fallback
        }
    }

    let initialIndexed: Wrapper?

    init(node: FNode, index: FIndex = FPriorityIndex.priorityIndex, indexed: Wrapper? = nil) {
        self.node = node
        self.index = index
        self.initialIndexed = indexed
    }

    static func indexedNode(node: FNode) -> FIndexedNode {
        indexedNodeWithNode(node, index: FPriorityIndex.priorityIndex)
    }

    static func indexedNodeWithNode(_ node: FNode, index: FIndex) -> FIndexedNode {
        .init(node: node, index: index)
    }

    func hasIndex(_ index: FIndex) -> Bool {
        self.index.isEqual(index)
    }

    func updateChild(_ key: String, withNewChild newChildNode: FNode) -> Self {
        let newNode = node.updateImmediateChild(key, withNewChild: newChildNode)
        switch indexed {
        case .fallback:
            if !self.index.isDefined(on: newChildNode) {
                // doesn't affect the index, no need to create an index
                return FIndexedNode(node: newNode, index: index, indexed: Wrapper.fallback)
            } else {
                // No need to index yet, index lazily
                #warning("Does this not actually produce the same as the above? Given that the fallback case has a zero cost?")
                return FIndexedNode(node: newNode, index: index)
            }
        case .indexed(var set):
            let oldChild = node.getImmediateChild(key)
            let oldEntry = IndexedNamedNode(wrapped: FNamedNode(name: key, andNode: oldChild), index: index)
            set.remove(oldEntry)
            if !newChildNode.isEmpty {
                let newEntry = IndexedNamedNode(wrapped: FNamedNode(name: key, andNode: newChildNode), index: index)
                set.insert(newEntry)
            }
            return FIndexedNode(node: newNode, index: index, indexed: .indexed(set))
        }

    }
    func updatePriority(_ priority: FNode) -> FIndexedNode {
        FIndexedNode(node: node.updatePriority(priority),
                     index: index,
                     indexed: indexed)
    }

    var firstChild: FNamedNode? {
        guard let childrenNode = node as? FChildrenNode else {
            return nil
        }
        switch indexed {
        case .fallback:
            return childrenNode.firstChild()
        case .indexed(let set):
            return set.first?.wrapped
        }
    }

    var lastChild: FNamedNode? {
        guard let childrenNode = node as? FChildrenNode else {
            return nil
        }
        switch indexed {
        case .fallback:
            return childrenNode.lastChild()
        case .indexed(let set):
            return set.last?.wrapped
        }
    }

    func predecessorForChildKey(_ childKey: String, childNode: FNode, index: FIndex) -> String? {
        if !self.index.isEqual(index) {
            fatalError("Index not available in IndexedNode!")
//            [NSException raise:NSInvalidArgumentException
//                        format:@"Index not available in IndexedNode!"];
        }
        switch indexed {
        case .fallback:
            return node.predecessorChildKey(childKey)
        case .indexed(let set):
            let entry = IndexedNamedNode(wrapped: FNamedNode(name: childKey, andNode: childNode), index: self.index)
            guard let index = set.firstIndex(of: entry), index != set.startIndex else {
                return nil
            }
            return set[set.index(before: index)].wrapped.name
        }
    }

    func enumerateChildrenReverse(_ reverse: Bool, usingBlock block: @escaping (String, FNode, UnsafeMutablePointer<ObjCBool>) -> Void) {
        switch indexed {
        case .fallback:
            node.enumerateChildrenReverse(reverse, usingBlock: block)
        case .indexed(let set):
            var stop = ObjCBool(booleanLiteral: false)
            if reverse {
                for key in set.reversed() {
                    block(key.wrapped.name, key.wrapped.node, &stop)
                    if stop.boolValue {
                        break
                    }
                }
            } else {
                for key in set {
                    block(key.wrapped.name, key.wrapped.node, &stop)
                    if stop.boolValue {
                        break
                    }
                }
            }
        }
    }

    var children: FIndexedNodeChildren {
        .init(node: self)
    }
}

struct FIndexedNodeChildren: Sequence {
    fileprivate let node: FIndexedNode
    fileprivate init(node: FIndexedNode) {
        self.node = node
    }
    func makeIterator() -> NamedNodeIterator {
        .init(node: node)
    }
}

struct NamedNodeIterator: IteratorProtocol {
    typealias Element = FNamedNode

    var _next: () -> FNamedNode?

    fileprivate init(node: FIndexedNode) {
        switch node.indexed {
        case .fallback:
            if let childrenNode = node.node as? FChildrenNode {
                var iterator = childrenNode.children.makeIterator()
                _next = {
                    guard let element = iterator.next() else { return nil }
                    return FNamedNode(name: element.key.key, andNode: element.value)
                }
            } else {
                _next = { nil }
            }
        case .indexed(let set):
            var iterator = set.makeIterator()
            _next = {
                guard let element = iterator.next() else { return nil }
                return element.wrapped
            }
        }
    }

    func next() -> FNamedNode? { _next() }
}

@objc(FIndexedNode) public class FIndexedNodeObjC: NSObject {
    internal let wrapped: FIndexedNode
    @objc public var node: FNode {
        wrapped.node
    }

    @objc public static func indexedNode(node: FNode) -> FIndexedNodeObjC {
        indexedNodeWithNode(node, index: FPriorityIndex.priorityIndex)
    }
    @objc public static func indexedNodeWithNode(_ node: FNode, index: FIndex) -> FIndexedNodeObjC {
        .init(node: node, index: index)
    }

    internal init(wrapped: FIndexedNode) {
        self.wrapped = wrapped
    }

    @objc public init(node: FNode) {
        self.wrapped = FIndexedNode(node: node)
    }

    @objc public init(node: FNode, index: FIndex) {
        self.wrapped = FIndexedNode(node: node, index: index)
    }

    @objc public func hasIndex(_ index: FIndex) -> Bool {
        wrapped.index.isEqual(index)
    }
    @objc public func updateChild(_ key: String, withNewChild newChildNode: FNode) -> FIndexedNodeObjC {
        let x = wrapped.updateChild(key, withNewChild: newChildNode)
        return FIndexedNodeObjC(wrapped: x)
    }
    @objc public func updatePriority(_ priority: FNode) -> FIndexedNodeObjC {
        let updated = wrapped.updatePriority(priority)
        return .init(wrapped: updated)
    }
    @objc public var firstChild: FNamedNode? {
        wrapped.firstChild
    }
    @objc public var lastChild: FNamedNode? {
        wrapped.lastChild
    }
    @objc public func predecessorForChildKey(_ childKey: String, childNode: FNode, index: FIndex) -> String? {
        wrapped.predecessorForChildKey(childKey, childNode: childNode, index: index)
    }
    @objc public func enumerateChildrenReverse(_ reverse: Bool, usingBlock block: @escaping (String, FNode, UnsafeMutablePointer<ObjCBool>) -> Void) {
        wrapped.enumerateChildrenReverse(reverse, usingBlock: block)
    }

    var children: FIndexedNodeChildren { wrapped.children }
}

//@objc public class FIndexedNode: NSObject {
//    @objc public func childEnumerator() -> NSEnumerator {
//        switch indexed {
//        case .fallback:
//            return node.childEnumerator()
//        case .indexed(let set):
//            return ObjectEnumerator(iterator: set.makeIterator())
//        }
//    }
//}

//class ObjectEnumerator: NSEnumerator {
//    var iterator: IndexingIterator<(OrderedSet<FNamedNode>)>
//    init(iterator: IndexingIterator<(OrderedSet<FNamedNode>)>) {
//        self.iterator = iterator
//    }
//
//    override func nextObject() -> Any? {
//        iterator.next()
//    }
//
//}
